sql: fix unescaped casts to regclass types

otan opened the pull request:
Release note (bug fix): Fix a bug where casts to
regclass were not escaped (i.e. when the
type or table name had " characters).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54947)
<!-- Reviewable:end -->

otan committed:
sql: fix unescaped casts to regclass types

Release note (bug fix): Fix a bug where casts to
regclass were not escaped (i.e. when the
type or table name had " characters).

otan commented:
oops, generalised https://github.com/cockroachdb/cockroach/pull/55607 to include these.
